homeassistant/components/modem_callerid/sensor.py

```diff
@@ -1,121 +1,126 @@
-    def native_value(self):
-        """Return the state of the device."""
-        return self._state
-
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        return self._name
-
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes."""
-        return self._attributes
-
-    def _stop_modem(self, event):
-        """HA is shutting down, close modem port."""
-        if self.modem:
-            self.modem.close()
-            self.modem = None
+    _attr_icon = ICON
+    _attr_should_poll = False
 
-    def _incomingcallcallback(self, newstate):
+    def __init__(
+        self, api: PhoneModem, name: str, device: str, server_unique_id: str
+    ) -> None:
+        """Initialize the sensor."""
+        self.device = device
+        self.api = api
+        self._attr_name = name
+        self._attr_unique_id = server_unique_id
+        self._attr_native_value = STATE_IDLE
+        self._attr_extra_state_attributes = {
+            CID.CID_TIME: 0,
+            CID.CID_NUMBER: "",
+            CID.CID_NAME: "",
+        }
+
+    async def async_added_to_hass(self) -> None:
+        """Call when the modem sensor is added to Home Assistant."""
+        self.api.registercallback(self._async_incoming_call)
+        await super().async_added_to_hass()
+
+    @callback
+    def _async_incoming_call(self, new_state) -> None:
         """Handle new states."""
-        if newstate == self.modem.STATE_RING:
-            if self.state == self.modem.STATE_IDLE:
-                att = {
-                    "cid_time": self.modem.get_cidtime,
-                    "cid_number": "",
-                    "cid_name": "",
+        if new_state == PhoneModem.STATE_RING:
+            if self.native_value == PhoneModem.STATE_IDLE:
+                self._attr_extra_state_attributes = {
+                    CID.CID_NUMBER: "",
+                    CID.CID_NAME: "",
                 }
-                self.set_attributes(att)
-            self._state = STATE_RING
-            self.schedule_update_ha_state()
-        elif newstate == self.modem.STATE_CALLERID:
-            att = {
-                "cid_time": self.modem.get_cidtime,
-                "cid_number": self.modem.get_cidnumber,
-                "cid_name": self.modem.get_cidname,
+        elif new_state == PhoneModem.STATE_CALLERID:
+            self._attr_extra_state_attributes = {
+                CID.CID_NUMBER: self.api.cid_number,
+                CID.CID_NAME: self.api.cid_name,
             }
-            self.set_attributes(att)
-            self._state = STATE_CALLERID
-            self.schedule_update_ha_state()
-        elif newstate == self.modem.STATE_IDLE:
-            self._state = STATE_IDLE
-            self.schedule_update_ha_state()
+        self._attr_extra_state_attributes[CID.CID_TIME] = self.api.cid_time
+        self._attr_native_value = self.api.state
+        self.async_schedule_update_ha_state()
```

Comment:
```suggestion
        self.async_write_ha_state()
```

If the callback is running in the event loop, you should be able to write state directly.